From 7f2ed125dde34d0fbbb0a6dcba4e1135738d396b Mon Sep 17 00:00:00 2001
From: Hardik T Shah <hardik.t.shah@intel.com>
Date: Thu, 10 Mar 2016 21:37:29 +0530
Subject: [PATCH 3862/4301] SoundWire:Intel: Add function to the clock
 frequency on bus.

commit 6593e330953e879481f8c5b622e0dbdcc755d562 from
git://git@github.com:01org/linux-intel-4.9.git

This patch adds the method to be used by bus driver to
set the bus frequency. This is used by bus driver to
increase or decrease the clock frequency based on
bus bandwidth requirement. Bus bandwidth is determined
based on active audio streams on bus.

Change-Id: Ifb2bfbeaea4cb81ae54e859b3ffa76015f8f394b
Signed-off-by: Hardik T Shah <hardik.t.shah@intel.com>
---
 drivers/sdw/sdw_cnl.c |   29 +++++++++++++++++++++++++++++
 1 files changed, 29 insertions(+), 0 deletions(-)

diff --git a/drivers/sdw/sdw_cnl.c b/drivers/sdw/sdw_cnl.c
index 517e86c..c5ec52f 100644
--- a/drivers/sdw/sdw_cnl.c
+++ b/drivers/sdw/sdw_cnl.c
@@ -503,6 +503,35 @@ static int cnl_sdw_set_ssp_interval(struct sdw_master *mstr,
 static int cnl_sdw_set_clock_freq(struct sdw_master *mstr,
 			int cur_clk_freq, int bank)
 {
+	struct cnl_sdw *sdw = sdw_master_get_drvdata(mstr);
+	struct cnl_sdw_data *data = &sdw->data;
+	int mcp_clockctrl_offset, mcp_clockctrl;
+
+
+	/* TODO: Retrieve divider value or get value directly from calling
+	 * function
+	 */
+	int divider = ((9600000/cur_clk_freq) - 1);
+
+	if (bank) {
+		mcp_clockctrl_offset = SDW_CNL_MCP_CLOCKCTRL1;
+		mcp_clockctrl = cnl_sdw_reg_readl(data->sdw_regs,
+				SDW_CNL_MCP_CLOCKCTRL1);
+
+	} else {
+		mcp_clockctrl_offset = SDW_CNL_MCP_CLOCKCTRL0;
+		mcp_clockctrl = cnl_sdw_reg_readl(data->sdw_regs,
+				SDW_CNL_MCP_CLOCKCTRL0);
+	}
+
+	mcp_clockctrl |= divider;
+
+	/* Write value here */
+	cnl_sdw_reg_writel(data->sdw_regs, mcp_clockctrl_offset,
+				mcp_clockctrl);
+
+	mcp_clockctrl = cnl_sdw_reg_readl(data->sdw_regs,
+				mcp_clockctrl_offset);
 	return 0;
 }
 
-- 
1.7.5.4

